### ISSUE TITLE
False positive on variable populated in exhaustive switch expression

### ISSUE THREAD

**Issue by wbadam** [uber/NullAway#1124]
Starting from version 0.11.0, NullAway reports false positive on a nullable variable after the switch expression:

```java
    public enum SomeEnum {
        A,
        B,
    }

    public int exhaustiveSwitch(SomeEnum e) {
        String s = null;
        switch (e) {
            case A -> {
                s = "A";
            }
            case B -> throw new RuntimeException();
        }
        return s.length(); // <-- warning: [NullAway] dereferenced expression s is @Nullable
    }
```

[Runnable example](https://github.com/wbadam/nullaway-java-sample/blob/adam.w-exhaustive-switch/lib/src/main/java/net/sridharan/sample/Library.java#L13-L27)

`s` at the `return` statement is non-nullable because the exhaustive switch either sets a value or makes the line unreachable.

Version 0.10.26 didn't report this warning.

**Comment by msridhar** [comment]
I can reproduce.  Thanks for the report.  I'll have to dig a bit deeper to understand what is going on here.

**Comment by msridhar** [comment]
After digging further, this issue has come up before, and was discussed on the Checker Framework issue tracker in https://github.com/typetools/checker-framework/issues/6644.  The key here is that in general, the enum could evolve with additional values, in which case the `switch` would no longer be exhaustive.  So, just looking at the code in `exhaustiveSwitch` it's actually not clear whether it is exhaustive or not.  Our analysis of `exhaustiveSwitch` does not take into account the fact that the declaration of the `enum` is in the same source file.  `javac` does not account for this either; see the example in https://github.com/typetools/checker-framework/issues/6644#issuecomment-2143832163, where if you eliminate the initializer on the variable the code no longer compiles.

Bottom line, while this is a false positive, we are probably not going to introduce special-case handling for it.  As workarounds, you can either add a `default` in the `switch`, or use a `switch` expression, for which a default case that throws an exception is automatically generated.

**Comment by msridhar** [comment]
It may also be useful to look at the generated bytecodes for some cases to see the difference between switch statements and expressions:

https://godbolt.org/z/GY59e6aYx

If you uncomment `exhaustiveSwitch2` you'll see a case that does not compile.
